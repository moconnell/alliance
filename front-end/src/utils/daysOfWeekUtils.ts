import DayOfWeek from "../types/dayOfWeek";
import DaysOfWeek from "../types/daysOfWeek";

export const toDaysOfWeek = (day: DayOfWeek) =>
  DaysOfWeek[DayOfWeek[day] as keyof typeof DaysOfWeek];

export const isAvailable = (date: Date, availableDays: DaysOfWeek) => {
  const day = date.getDay();
  const daysOfWeek = 1 << day;
  
  return (daysOfWeek & availableDays) === daysOfWeek;
};
