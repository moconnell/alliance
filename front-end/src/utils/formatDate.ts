import Time from "../types/time";

const locale =
  process.env.REACT_LOCALE ?? Intl.DateTimeFormat().resolvedOptions().locale;

const dateTimeFormat = Intl.DateTimeFormat(locale, {
  dateStyle: "full",
  timeStyle: "short",
});

const shortTimeFormat = Intl.DateTimeFormat(locale, {
  timeStyle: "short",
});

function isTime(value: Time | Date): value is Time {
  return (value as Time).hours !== undefined;
}

const toDate = (t: Time) => {
  let d = new Date();
  d.setHours(t.hours, t.minutes);
  return d;
};

export const formatTime = (value: Time | Date) => {
  return shortTimeFormat.format(isTime(value) ? toDate(value) : value);
};

export const formatDateTime = (value: Date) => {
  return dateTimeFormat.format(value);
};
