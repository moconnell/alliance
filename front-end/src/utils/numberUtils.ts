export const toBooleanArray = (binary: number, length: number = 0) =>
  Math.abs(binary)
    .toString(2)
    .padStart(length, "0")
    .split("")
    .map(Number)
    .map(Boolean);
