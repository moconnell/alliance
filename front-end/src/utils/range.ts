export function range(start: number, end: number): number[] {
  start = Math.floor(start);
  end = Math.floor(end);

  const diff = end - start;
  if (diff === 0) {
    return [start];
  }

  const keys = Array(Math.abs(diff) + 1).keys();
  return Array.from(keys).map((x) => {
    const increment = end > start ? x : -x;
    return start + increment;
  });
}
