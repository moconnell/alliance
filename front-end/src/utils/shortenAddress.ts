export function shortenAddress(address?: string) {
  if (!address) return address;
  const length = address.length;
  return `${address.slice(0, 5)}..${address.slice(length - 4, length)}`;
}
