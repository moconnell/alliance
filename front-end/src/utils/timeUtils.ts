import Time from "../types/time";
import assert from "assert";

export const totalMinutes = (t: Time) => t.hours * 60 + t.minutes;

export const totalSeconds = (t: Time) =>
  t.hours * 60 * 60 + t.minutes * 60 + (t.seconds ?? 0);

export const fromTotalMins = (mins: number) => {
  const hours = Math.floor(mins / 60) % 24;
  const minutes = mins % 60;
  assert(hours < 24);
  assert(minutes < 60);
  return { hours, minutes } as Time;
};

export const assertValid = (t?: Time) => {
  assert(t, "time undefined");
  assert(t.hours >= 0 && t.hours < 24, "invalid hours");
  assert(t.minutes >= 0 && t.minutes < 60, "invalid minutes");
};

export const compare = (t1: Time, t2: Time) => {
  const s1 = totalSeconds(t1);
  const s2 = totalSeconds(t2);
  if (s1 === s2) return 0;
  if (s1 < s2) return -1;
  return 1;
};

export const tomorrow = () => {
  let d = new Date();
  d.setDate(d.getDate() + 1);
  d.setHours(0, 0, 0, 0);
  return d;
};
