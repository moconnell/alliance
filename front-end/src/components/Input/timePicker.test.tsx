import { render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Time from "../../types/time";
import { formatTime } from "../../utils/formatDate";
import TimePicker from "./timePicker";

describe("<TimePicker />", () => {
  const testData: [string, string?, number?, Time?][] = [
    ["render correctly"],
    ["render prefix, interval correctly", "f", 15],
    [
      "render value correctly",
      undefined,
      undefined,
      { hours: 12, minutes: 30 },
    ],
  ];

  testData.forEach(([name, prefix, timeIntervalMins, value]) =>
    it(`should ${name}`, async () => {
      const { asFragment, getByTestId } = render(
        <TimePicker
          name="from"
          prefix={prefix}
          timeIntervalMins={timeIntervalMins}
          value={value}
        />
      );
      await waitFor(() =>
        expect(getByTestId("select:from")).toHaveDisplayValue(
          formatTime(value ?? { hours: 0, minutes: 0 })
        )
      );
      expect(asFragment()).toMatchSnapshot();
    })
  );

  it("should pass value on change", async () => {
    const user = userEvent.setup();
    const mockHandler = jest.fn();

    const { getByTestId } = render(
      <TimePicker name="from" onChange={mockHandler} />
    );

    await user.selectOptions(getByTestId("select:from"), ["12:30 PM"]);

    expect(mockHandler).toBeCalledWith({ hours: 12, minutes: 30 });
  });

  it("should handle missing onchange handler", async () => {
    const user = userEvent.setup();
    const mockHandler = jest.fn();

    const { getByTestId } = render(<TimePicker name="from" />);

    await user.selectOptions(getByTestId("select:from"), ["12:30 PM"]);

    expect(mockHandler).not.toBeCalledWith({ hours: 12, minutes: 30 });
  });
});
