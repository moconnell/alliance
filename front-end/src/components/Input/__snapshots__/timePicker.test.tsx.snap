// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimePicker /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-select__wrapper css-42b2qy"
  >
    <select
      class="chakra-select css-1ik61og"
      data-testid="select:from"
      name="from"
    >
      <option
        value="0"
      >
        12:00 AM
      </option>
      <option
        value="30"
      >
        12:30 AM
      </option>
      <option
        value="60"
      >
        1:00 AM
      </option>
      <option
        value="90"
      >
        1:30 AM
      </option>
      <option
        value="120"
      >
        2:00 AM
      </option>
      <option
        value="150"
      >
        2:30 AM
      </option>
      <option
        value="180"
      >
        3:00 AM
      </option>
      <option
        value="210"
      >
        3:30 AM
      </option>
      <option
        value="240"
      >
        4:00 AM
      </option>
      <option
        value="270"
      >
        4:30 AM
      </option>
      <option
        value="300"
      >
        5:00 AM
      </option>
      <option
        value="330"
      >
        5:30 AM
      </option>
      <option
        value="360"
      >
        6:00 AM
      </option>
      <option
        value="390"
      >
        6:30 AM
      </option>
      <option
        value="420"
      >
        7:00 AM
      </option>
      <option
        value="450"
      >
        7:30 AM
      </option>
      <option
        value="480"
      >
        8:00 AM
      </option>
      <option
        value="510"
      >
        8:30 AM
      </option>
      <option
        value="540"
      >
        9:00 AM
      </option>
      <option
        value="570"
      >
        9:30 AM
      </option>
      <option
        value="600"
      >
        10:00 AM
      </option>
      <option
        value="630"
      >
        10:30 AM
      </option>
      <option
        value="660"
      >
        11:00 AM
      </option>
      <option
        value="690"
      >
        11:30 AM
      </option>
      <option
        value="720"
      >
        12:00 PM
      </option>
      <option
        value="750"
      >
        12:30 PM
      </option>
      <option
        value="780"
      >
        1:00 PM
      </option>
      <option
        value="810"
      >
        1:30 PM
      </option>
      <option
        value="840"
      >
        2:00 PM
      </option>
      <option
        value="870"
      >
        2:30 PM
      </option>
      <option
        value="900"
      >
        3:00 PM
      </option>
      <option
        value="930"
      >
        3:30 PM
      </option>
      <option
        value="960"
      >
        4:00 PM
      </option>
      <option
        value="990"
      >
        4:30 PM
      </option>
      <option
        value="1020"
      >
        5:00 PM
      </option>
      <option
        value="1050"
      >
        5:30 PM
      </option>
      <option
        value="1080"
      >
        6:00 PM
      </option>
      <option
        value="1110"
      >
        6:30 PM
      </option>
      <option
        value="1140"
      >
        7:00 PM
      </option>
      <option
        value="1170"
      >
        7:30 PM
      </option>
      <option
        value="1200"
      >
        8:00 PM
      </option>
      <option
        value="1230"
      >
        8:30 PM
      </option>
      <option
        value="1260"
      >
        9:00 PM
      </option>
      <option
        value="1290"
      >
        9:30 PM
      </option>
      <option
        value="1320"
      >
        10:00 PM
      </option>
      <option
        value="1350"
      >
        10:30 PM
      </option>
      <option
        value="1380"
      >
        11:00 PM
      </option>
      <option
        value="1410"
      >
        11:30 PM
      </option>
    </select>
    <div
      class="chakra-select__icon-wrapper css-162mkon"
    >
      <svg
        aria-hidden="true"
        class="chakra-select__icon"
        focusable="false"
        role="presentation"
        style="width: 1em; height: 1em; color: currentColor;"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TimePicker /> should render prefix, interval correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-select__wrapper css-42b2qy"
  >
    <select
      class="chakra-select css-1ik61og"
      data-testid="select:from"
      name="from"
    >
      <option
        value="0"
      >
        12:00 AM
      </option>
      <option
        value="15"
      >
        12:15 AM
      </option>
      <option
        value="30"
      >
        12:30 AM
      </option>
      <option
        value="45"
      >
        12:45 AM
      </option>
      <option
        value="60"
      >
        1:00 AM
      </option>
      <option
        value="75"
      >
        1:15 AM
      </option>
      <option
        value="90"
      >
        1:30 AM
      </option>
      <option
        value="105"
      >
        1:45 AM
      </option>
      <option
        value="120"
      >
        2:00 AM
      </option>
      <option
        value="135"
      >
        2:15 AM
      </option>
      <option
        value="150"
      >
        2:30 AM
      </option>
      <option
        value="165"
      >
        2:45 AM
      </option>
      <option
        value="180"
      >
        3:00 AM
      </option>
      <option
        value="195"
      >
        3:15 AM
      </option>
      <option
        value="210"
      >
        3:30 AM
      </option>
      <option
        value="225"
      >
        3:45 AM
      </option>
      <option
        value="240"
      >
        4:00 AM
      </option>
      <option
        value="255"
      >
        4:15 AM
      </option>
      <option
        value="270"
      >
        4:30 AM
      </option>
      <option
        value="285"
      >
        4:45 AM
      </option>
      <option
        value="300"
      >
        5:00 AM
      </option>
      <option
        value="315"
      >
        5:15 AM
      </option>
      <option
        value="330"
      >
        5:30 AM
      </option>
      <option
        value="345"
      >
        5:45 AM
      </option>
      <option
        value="360"
      >
        6:00 AM
      </option>
      <option
        value="375"
      >
        6:15 AM
      </option>
      <option
        value="390"
      >
        6:30 AM
      </option>
      <option
        value="405"
      >
        6:45 AM
      </option>
      <option
        value="420"
      >
        7:00 AM
      </option>
      <option
        value="435"
      >
        7:15 AM
      </option>
      <option
        value="450"
      >
        7:30 AM
      </option>
      <option
        value="465"
      >
        7:45 AM
      </option>
      <option
        value="480"
      >
        8:00 AM
      </option>
      <option
        value="495"
      >
        8:15 AM
      </option>
      <option
        value="510"
      >
        8:30 AM
      </option>
      <option
        value="525"
      >
        8:45 AM
      </option>
      <option
        value="540"
      >
        9:00 AM
      </option>
      <option
        value="555"
      >
        9:15 AM
      </option>
      <option
        value="570"
      >
        9:30 AM
      </option>
      <option
        value="585"
      >
        9:45 AM
      </option>
      <option
        value="600"
      >
        10:00 AM
      </option>
      <option
        value="615"
      >
        10:15 AM
      </option>
      <option
        value="630"
      >
        10:30 AM
      </option>
      <option
        value="645"
      >
        10:45 AM
      </option>
      <option
        value="660"
      >
        11:00 AM
      </option>
      <option
        value="675"
      >
        11:15 AM
      </option>
      <option
        value="690"
      >
        11:30 AM
      </option>
      <option
        value="705"
      >
        11:45 AM
      </option>
      <option
        value="720"
      >
        12:00 PM
      </option>
      <option
        value="735"
      >
        12:15 PM
      </option>
      <option
        value="750"
      >
        12:30 PM
      </option>
      <option
        value="765"
      >
        12:45 PM
      </option>
      <option
        value="780"
      >
        1:00 PM
      </option>
      <option
        value="795"
      >
        1:15 PM
      </option>
      <option
        value="810"
      >
        1:30 PM
      </option>
      <option
        value="825"
      >
        1:45 PM
      </option>
      <option
        value="840"
      >
        2:00 PM
      </option>
      <option
        value="855"
      >
        2:15 PM
      </option>
      <option
        value="870"
      >
        2:30 PM
      </option>
      <option
        value="885"
      >
        2:45 PM
      </option>
      <option
        value="900"
      >
        3:00 PM
      </option>
      <option
        value="915"
      >
        3:15 PM
      </option>
      <option
        value="930"
      >
        3:30 PM
      </option>
      <option
        value="945"
      >
        3:45 PM
      </option>
      <option
        value="960"
      >
        4:00 PM
      </option>
      <option
        value="975"
      >
        4:15 PM
      </option>
      <option
        value="990"
      >
        4:30 PM
      </option>
      <option
        value="1005"
      >
        4:45 PM
      </option>
      <option
        value="1020"
      >
        5:00 PM
      </option>
      <option
        value="1035"
      >
        5:15 PM
      </option>
      <option
        value="1050"
      >
        5:30 PM
      </option>
      <option
        value="1065"
      >
        5:45 PM
      </option>
      <option
        value="1080"
      >
        6:00 PM
      </option>
      <option
        value="1095"
      >
        6:15 PM
      </option>
      <option
        value="1110"
      >
        6:30 PM
      </option>
      <option
        value="1125"
      >
        6:45 PM
      </option>
      <option
        value="1140"
      >
        7:00 PM
      </option>
      <option
        value="1155"
      >
        7:15 PM
      </option>
      <option
        value="1170"
      >
        7:30 PM
      </option>
      <option
        value="1185"
      >
        7:45 PM
      </option>
      <option
        value="1200"
      >
        8:00 PM
      </option>
      <option
        value="1215"
      >
        8:15 PM
      </option>
      <option
        value="1230"
      >
        8:30 PM
      </option>
      <option
        value="1245"
      >
        8:45 PM
      </option>
      <option
        value="1260"
      >
        9:00 PM
      </option>
      <option
        value="1275"
      >
        9:15 PM
      </option>
      <option
        value="1290"
      >
        9:30 PM
      </option>
      <option
        value="1305"
      >
        9:45 PM
      </option>
      <option
        value="1320"
      >
        10:00 PM
      </option>
      <option
        value="1335"
      >
        10:15 PM
      </option>
      <option
        value="1350"
      >
        10:30 PM
      </option>
      <option
        value="1365"
      >
        10:45 PM
      </option>
      <option
        value="1380"
      >
        11:00 PM
      </option>
      <option
        value="1395"
      >
        11:15 PM
      </option>
      <option
        value="1410"
      >
        11:30 PM
      </option>
      <option
        value="1425"
      >
        11:45 PM
      </option>
    </select>
    <div
      class="chakra-select__icon-wrapper css-162mkon"
    >
      <svg
        aria-hidden="true"
        class="chakra-select__icon"
        focusable="false"
        role="presentation"
        style="width: 1em; height: 1em; color: currentColor;"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TimePicker /> should render value correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-select__wrapper css-42b2qy"
  >
    <select
      class="chakra-select css-1ik61og"
      data-testid="select:from"
      name="from"
    >
      <option
        value="0"
      >
        12:00 AM
      </option>
      <option
        value="30"
      >
        12:30 AM
      </option>
      <option
        value="60"
      >
        1:00 AM
      </option>
      <option
        value="90"
      >
        1:30 AM
      </option>
      <option
        value="120"
      >
        2:00 AM
      </option>
      <option
        value="150"
      >
        2:30 AM
      </option>
      <option
        value="180"
      >
        3:00 AM
      </option>
      <option
        value="210"
      >
        3:30 AM
      </option>
      <option
        value="240"
      >
        4:00 AM
      </option>
      <option
        value="270"
      >
        4:30 AM
      </option>
      <option
        value="300"
      >
        5:00 AM
      </option>
      <option
        value="330"
      >
        5:30 AM
      </option>
      <option
        value="360"
      >
        6:00 AM
      </option>
      <option
        value="390"
      >
        6:30 AM
      </option>
      <option
        value="420"
      >
        7:00 AM
      </option>
      <option
        value="450"
      >
        7:30 AM
      </option>
      <option
        value="480"
      >
        8:00 AM
      </option>
      <option
        value="510"
      >
        8:30 AM
      </option>
      <option
        value="540"
      >
        9:00 AM
      </option>
      <option
        value="570"
      >
        9:30 AM
      </option>
      <option
        value="600"
      >
        10:00 AM
      </option>
      <option
        value="630"
      >
        10:30 AM
      </option>
      <option
        value="660"
      >
        11:00 AM
      </option>
      <option
        value="690"
      >
        11:30 AM
      </option>
      <option
        value="720"
      >
        12:00 PM
      </option>
      <option
        value="750"
      >
        12:30 PM
      </option>
      <option
        value="780"
      >
        1:00 PM
      </option>
      <option
        value="810"
      >
        1:30 PM
      </option>
      <option
        value="840"
      >
        2:00 PM
      </option>
      <option
        value="870"
      >
        2:30 PM
      </option>
      <option
        value="900"
      >
        3:00 PM
      </option>
      <option
        value="930"
      >
        3:30 PM
      </option>
      <option
        value="960"
      >
        4:00 PM
      </option>
      <option
        value="990"
      >
        4:30 PM
      </option>
      <option
        value="1020"
      >
        5:00 PM
      </option>
      <option
        value="1050"
      >
        5:30 PM
      </option>
      <option
        value="1080"
      >
        6:00 PM
      </option>
      <option
        value="1110"
      >
        6:30 PM
      </option>
      <option
        value="1140"
      >
        7:00 PM
      </option>
      <option
        value="1170"
      >
        7:30 PM
      </option>
      <option
        value="1200"
      >
        8:00 PM
      </option>
      <option
        value="1230"
      >
        8:30 PM
      </option>
      <option
        value="1260"
      >
        9:00 PM
      </option>
      <option
        value="1290"
      >
        9:30 PM
      </option>
      <option
        value="1320"
      >
        10:00 PM
      </option>
      <option
        value="1350"
      >
        10:30 PM
      </option>
      <option
        value="1380"
      >
        11:00 PM
      </option>
      <option
        value="1410"
      >
        11:30 PM
      </option>
    </select>
    <div
      class="chakra-select__icon-wrapper css-162mkon"
    >
      <svg
        aria-hidden="true"
        class="chakra-select__icon"
        focusable="false"
        role="presentation"
        style="width: 1em; height: 1em; color: currentColor;"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
